### ISSUE TITLE
Autoinstallation support for FreeBSD

### ISSUE THREAD

**Issue by SchoolGuy** [cobbler/cobbler#3280]
### Is your feature request related to a problem?

Currently our FreeBSD support is limited to manual installations. IIf many systems are installed or the same system very often this is a time consuming task that doesn't need to be performed.

### Provide a detailed description of the proposed feature

Unattended auto-installation support for FreeBSD.

### Alternatives you've considered

Leaving as is: Well our strength is auto-installation, so let's use it.

### Additional information

I found some pointers already:

- https://forums.freebsd.org/threads/freebsd-and-automated-pxe-setup.81712/
- https://www.freebsd.org/cgi/man.cgi?query=bsdinstall&sektion=8&manpath=freebsd-release-ports
- https://docs.freebsd.org/en/articles/remote-install/
- https://www.freebsd.org/cgi/man.cgi?query=bsdconfig&sektion=8
- https://docs.freebsd.org/doc/7.4-RELEASE/usr/share/doc/de/books/handbook/using-sysinstall.html
- https://www.youtube.com/watch?v=7F3UwfNB2JA

**Comment by Obi-LAN** [comment]
Hi @SchoolGuy :) I was sort of able to get this working using NFS instead of HTTP; unattended FreeBSD 13.0 PXE boot+install (as well as TrueNAS Core 13, and hopefully pfSense!), Cobbler v3.3.6 running on Fedora 34 Workstation. Thank you for all your help maintaining and enhancing the project!

I'm new to GitHub (but not Cobbler, I started with v2.8.5 in 2019) and these are only some suggestions, but I'm happy to contribute to any of the code or documentation to improve the project. No rush, but I would appreciate some feedback, or let's sync up on these findings. 

A few obstacles getting there:

1.) You mentioned freebsd support is limited to manual install, but I couldn't even get that working without most of the below fixes, so perhaps I'm going in the complete wrong direction. Unfortunately, this compatibility matrix (https://cobbler.readthedocs.io/en/v2.8.5/1_about/distribution_support.html) does not exist in the current documentation (or this guide https://cobbler.readthedocs.io/en/v2.8.5/1_about/distribution_notes.html#freebsd), so it was difficult to tell whether it was officially supported in any capacity, but I figured if "import" was supported in that version, and the new versions of FreeBSD are continuously being added to the Cobbler "distro-signatures.json" file, an import should still be possible with the latest release, I didn't want to use the "cobbler image" method since that seemed more complicated, and there was a bug in the latest docs (https://cobbler.readthedocs.io/en/latest/cobbler.html#memdisk-oracle-sun-maintenance-cs) reporting the "memdisk" type as being unusable, which might be correct for BSD official "memstick" installation media.

2.) some dependencies necessary were missing from the Cobbler docs, namely "python-magic" which I found in the "Fedora34.dockerfile" in Cobbler's source code. 

3.) The current distro signatures for the FreeBSD breed lists "base.txz" as the kernel and initrd, but the "utils.py" cobbler script does not include those in the "_re_kernel" or "_re_initrd" object/variable. I'm not sure if utils.py is a static script, or dynamically generated/configurable (perhaps through "import_rsync_whitelist" settings file, another conflicting/confusing source of info regarding supported breeds) but Cobbler successfully imported the kernel and initrd after a small edit to that script 

4.) another problem with the import process was the "version_file" field listed in "distro-signatures.json" for BSD 13 as "freebsd-version" which has a MIME type detected by magic/file as "text/x-shellscript" which is not included in Cobbler's "import_signatures.py" - it should be treated the same as "text/plain" is in that script in order to properly detect the version

5.) finally (and this took some debugging, a first for me) the line in "import_signatures.py" reading "if f_re.match(fname)" should instead read "if f_re.fullmatch(fname)" so that the "version_file" field will require an exact match for the version file name (all of the regex in the "distro_signatures.json" file looks quite precise and complete, so I would assume "fullmatch()" would be the right choice in most cases, but it is more strict). Otherwise Cobbler will detect the /usr/share/man/man5/device.hints.5.gz file on the FreeBSD 13 ISO thinking it is an older version, resulting in a "cannot use a string pattern on a bytes-like object" Python error. I corrected this with the 1 line change above to finally be able to import BSD 13. 2 concerns are that the comment in the "scan_signatures()" function in import_signatures.py says it looks for a match on signature file AND directory. If the "signature directory" refers to the "signatures" object in "distro_signatures.json" (I believe it does) then the import should not detect any of those files, since they are not in the listed "boot" directory for BSD breeds; If the import is only supposed to look in that directory first, then the comment should be clarified.

6.) I could not get FreeBSD to network boot using pxelinux, iPXE,or GRUB, only with the native "pxeboot" bootloader included with the official FreeBSD 13.0 ISO. After importing the distro, I was able to add a Cobbler System and set the "--filename" option to "pxeboot" I then copied the file to the "/var/lib/cobbler/loaders" directory (should this be achievable with the cobbler distro --boot-files option?), ran a cobbler sync, and was able to boot the file over the network, however, it would just hang indefinitely with no output without the next step.

7.) I didn't want to accept it, but as noted above, as well as in the official FreeBSD guide (https://docs.freebsd.org/en/books/handbook/advanced-networking/#network-diskless), I had to use the "nfs-server" package which came installed with Fedora 34 Workstation. Using NFS instead of HTTP means Cobbler does not detect that the system is booting (or possibly something to do with using BSD's native PXE bootloader?) leaving it incompatible with Cobbler --netboot-enabled/pxe_just_once, and other options/features. HOWEVER with the "distro_mirror/BSD-13" directory shared as an NFS export, and one additional, manual DHCP change (could probably be accomplished with a sync trigger) I was able to load and install FreeBSD 13 over the network.

8.) delete the contents of the imported "/etc/fstab" to bypass the CD-ROM mounting error 

As for unattended installation, just configure this 1 file in the imported tree: "/etc/installerconfig" to your specifications (more info here: https://punkt.de/de/blog/2017/automatisierte-installation-von-servern-mit-freebsd-und-zfs.html).

I figured I would start a conversation here before submitting a PR/commit, sorry for brain dumping. Overall, I think basic support for this through "modules.conf" file would be doable, at least very basic support, but I guess "--available-as" import option may be the intended alternative. I'm happy to provide my full list of commands from Fedora base installation to unattended BSD installation, including source code fixes, downloads, and system settings.

Sidenote: I could not seem to get some of the logging working in import_signatures.py script, namely the "Found a candidate signature" entry, even after enabling all debug level logging in logging_config.conf

**Comment by Obi-LAN** [comment]
perhaps there is a way to boot FreeBSD natively using only TFTP, as noted here https://man.freebsd.org/cgi/man.cgi?query=pxeboot&sektion=8

I will look into this this week hopefully.

**Comment by SchoolGuy** [comment]
@sus-admin I just saw your blog about Cobbler and thank you for putting in the work. In case you have any PRs for the work that you have started, I will gladly collaborate with you on getting it in a mergable state. I personally have no use for it but I am sure that other community members will be thankful for support on this matter.

**Comment by sus-admin** [comment]
@SchoolGuy I would love to get some of those blogs merged into the cobbler.github.io project. I will get a PR up for my "3.3.6 Beginners Guide" and notify you here as well once ready. Once that is merged I will submit more PRs for the additional guides I have created.

**Comment by sus-admin** [comment]
@SchoolGuy I've submitted a PR for my Cobbler beginner's guide, let me know your thoughts!
https://github.com/cobbler/cobbler.github.io/pull/186
